pinot-core/src/main/java/com/linkedin/pinot/core/query/pruner/PartitionSegmentPruner.java

```diff
@@ -99,8 +103,41 @@ public boolean pruneSegment(FilterQueryTree filterQueryTree, Map<String, ColumnM
       return false;
     }
 
-    Comparable value = getValue(filterQueryTree.getValue().get(0), metadata.getDataType());
     PartitionFunction partitionFunction = metadata.getPartitionFunction();
+
+    if (operator == FilterOperator.EQUALITY) {
+      return pruneEquality(filterQueryTree, metadata, partitionRanges, partitionFunction);
+    }
+
+    return pruneIn(filterQueryTree, metadata, partitionRanges, partitionFunction);
+  }
+
+  private boolean pruneIn(FilterQueryTree filterQueryTree, ColumnMetadata metadata, List<IntRange> partitionRanges,
+      PartitionFunction partitionFunction) {
+    Splitter valueSplitter = Splitter.on(IN_CLAUSE_DELIMITER);
+
+    List<String> values = valueSplitter.splitToList(filterQueryTree.getValue().get(0));
+
+    List<Integer> partitions;
+    partitions = new ArrayList<Integer>();
+
+    for (String value : values) {
```

Comment:
This is inefficient. We don't need to get all values upfront. That way we can have early bailout.